agent/lib/queue.js

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2017 Red Hat Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+var create_podgroup = require('./podgroup.js');
+var kubernetes = require('../lib/kubernetes.js');
+var log = require('./log.js').logger();
+var pod_watcher = require('./pod_watcher');
+
+function Queue (name) {
+    this.name = name;
+    this.pods  = create_podgroup();
+};
+
+function get_current_pods () {
+
+    var options = {
+        selector: 'role=broker'
+    };
+
+    if (process.env.INFRA_UUID) {
+        var infraSelector = "infraUuid=" + process.env.INFRA_UUID;
+        if (options.selector) {
+            options.selector += "," + infraSelector;
+        }
+    }
+    return kubernetes.get('pods', options);
+}
+
+function find_queue (name, pods) {
+    return Promise.all(pods.map(function (pod) {
+        return pod.broker.getQueueNames();
+    } )).then(function (results) {
+            var found = false;
+            var pod = undefined;
+            var min = undefined;
+            for (var i = 0; !found && i < results.length; i++) {
+                if (results[i].indexOf(name) >= 0) {
+                    pod = pods[i];
+                    found = true;
+                } else if (min === undefined || min > results[i].length) {
+                    pod = pods[i];
+                }
+            }
+            return {'found':found, 'pod':pod};
+        }
+    ).catch(function(e) {
+        log.error(e);
+    });
+};
+
+Queue.prototype.purge = function () {
+    var queueName = this.name;
+    var pods = this.pods;
+    return get_current_pods().then(
+        function(currentPods) {
+
+            var podDefs = currentPods.items.map(function (pod) {
+                return pod_watcher.get_pod_definition(pod);
+            });
+            pods.update(podDefs);
+
+            return find_queue(queueName, pods.pod_list()).then(
+                function (result) {
+
```

Comment:
what if a queue is shared across many brokers?  we need to purge each shard.

Reply:
using the address definition status, to get the brokers for standard address space.  If there are shards, it gets all of them.